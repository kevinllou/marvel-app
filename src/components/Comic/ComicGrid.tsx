/* eslint-disable react/no-unescaped-entities */
import { useSearchParams } from 'react-router-dom';
import useFetch from '../../hooks/useFetch';
import IApiResponse from '../../interfaces/IApiResponse';
import IComics from '../../interfaces/IComics';
import Spinner from '../Spinner/Spinner';
import ComicCard from './ComicCard';
import createURLFetch from '../../helpers/createURLFetch';
import IURLParams from '../../interfaces/IURLParams';
/* import usePagination from '../../hooks/usePagination';
import Pagination from '../Pagination/Pagination'; */
function ComicGrid() {
  const [searchParams] = useSearchParams();
  const comicsFetch: IURLParams = {
    titleParam: searchParams.get('title') || '',
    formatParam: searchParams.get('format') || '',
    page: searchParams.get('page') || '1',
  };
  const { state, data, error } = useFetch<IApiResponse<IComics>>(createURLFetch(comicsFetch, 'comics'));
  /*   const paginationRange = usePagination(
    {
      totalCount: data?.data?.total,
      pageSize: LIMIT,
      currentPage: page,
    },
  ); */
  if (state === 'loading') return <Spinner />;
  if (error) return <p>There was an error</p>;

  return (
    <>
      <div className="characters__grid">
        {data?.data?.results?.length === 0 && <p style={{ paddingTop: '.5rem' }}>We couldn't find any data ...</p>}
        {data?.data?.results?.map(({ id, title, thumbnail }) => (
          <ComicCard
            key={id}
            title={title}
            thumbnail={thumbnail}
            id={id}
          />
        ))}
      </div>
      {/*   <Pagination currentPage={page} paginationRange={paginationRange} /> */}
    </>
  );
}

export default ComicGrid;
