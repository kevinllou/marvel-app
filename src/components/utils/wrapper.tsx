import { render } from '@testing-library/react';
import { ReactElement } from 'react';
import { BrowserRouter, MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { store } from '../../redux/store';
import MarvelRoutes from '../../routes/MarvelRoutes';

const renderWithBrowserRouter = (component: ReactElement) => render(
  <Provider store={store}>
    <BrowserRouter>{component}</BrowserRouter>
  </Provider>,
);

const renderWithMemoryRouter = (path: string[] = ['/']) => render(
  <Provider store={store}>
    <MemoryRouter initialEntries={path}>
      <MarvelRoutes />
    </MemoryRouter>
  </Provider>,
);

export {
  renderWithBrowserRouter,
  renderWithMemoryRouter,
};
