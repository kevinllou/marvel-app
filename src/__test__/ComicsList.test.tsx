import { waitFor } from '@testing-library/react';
import { renderWithBrowserRouter } from '../components/utils/wrapper';
import ComicGrid from '../components/Comic/ComicGrid';

describe('Characters List Component', () => {
  beforeEach(() => {
    jest.spyOn(console, 'warn').mockImplementation(() => {});
    jest.spyOn(console, 'error').mockImplementation(() => {});
  });

  it('should render list of comics', async () => {
    const { container, getByText } = renderWithBrowserRouter(<ComicGrid />);
    await waitFor(() => {
      expect(container.querySelector('.spinner')).not.toBeTruthy();
      expect(getByText(/Marvel Previews/i)).toBeInTheDocument();
      expect(container).toMatchSnapshot();
    });
  });
});
