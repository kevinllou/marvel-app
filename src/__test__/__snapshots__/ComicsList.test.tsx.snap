// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Characters List Component should render list of comics 1`] = `
<div>
  <div
    class="characters__grid"
  >
    <div
      class="comics__gridItems"
    >
      <div
        class="comics__card"
      >
        <div
          class="comics__cardImage"
        >
          <img
            alt=""
            src="http://i.annihil.us/u/prod/marvel/i/mg/b/40/image_not_available.jpg"
          />
        </div>
        <div
          class="comics__cardBody"
        >
          <div
            class="comics__name"
          >
            <a
              href="/comics/82967"
            >
              Marvel Previews (2017)
            </a>
          </div>
          <div
            class="comics__options"
          >
            <button
              style="color: white;"
              type="button"
            >
              <i
                class="fa-sharp fa-solid fa-heart"
              />
            </button>
            <button
              style="color: white; pointer-events: auto;"
              type="button"
            >
              <i
                class="fa-solid fa-eye-slash"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="comics__gridItems"
    >
      <div
        class="comics__card"
      >
        <div
          class="comics__cardImage"
        >
          <img
            alt=""
            src="http://i.annihil.us/u/prod/marvel/i/mg/b/40/image_not_available.jpg"
          />
        </div>
        <div
          class="comics__cardBody"
        >
          <div
            class="comics__name"
          >
            <a
              href="/comics/82965"
            >
              Marvel (2017)
            </a>
          </div>
          <div
            class="comics__options"
          >
            <button
              style="color: white;"
              type="button"
            >
              <i
                class="fa-sharp fa-solid fa-heart"
              />
            </button>
            <button
              style="color: white; pointer-events: auto;"
              type="button"
            >
              <i
                class="fa-solid fa-eye-slash"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
  <section
    class="pagination"
  >
    <button
      class="pagination__prev"
      disabled=""
      type="button"
    >
      <i
        class="fa-solid fa-angle-left"
      />
    </button>
    <button
      class="pagination__btn pagination__btn--active"
      style="pointer-events: auto;"
      type="button"
    >
      1
    </button>
    <button
      class="pagination__btn"
      style="pointer-events: auto;"
      type="button"
    >
      2
    </button>
    <button
      class="pagination__btn"
      style="pointer-events: auto;"
      type="button"
    >
      3
    </button>
    <button
      class="pagination__btn"
      style="pointer-events: auto;"
      type="button"
    >
      4
    </button>
    <button
      class="pagination__btn"
      style="pointer-events: auto;"
      type="button"
    >
      5
    </button>
    <button
      class="pagination__btn"
      style="pointer-events: none;"
      type="button"
    >
      ...
    </button>
    <button
      class="pagination__btn"
      style="pointer-events: auto;"
      type="button"
    >
      78
    </button>
    <button
      class="pagination__next"
      type="button"
    >
      <i
        class="fa-solid fa-angle-right"
      />
    </button>
  </section>
</div>
`;
